File: src/com/intellij/advancedExpressionFolding/processor/expression/LiteralExpressionExt.java
package com.intellij.advancedExpressionFolding.processor.expression;

import com.intellij.advancedExpressionFolding.expression.Expression;
import com.intellij.advancedExpressionFolding.expression.literal.CharacterLiteral;
import com.intellij.advancedExpressionFolding.expression.literal.NumberLiteral;
import com.intellij.advancedExpressionFolding.expression.literal.StringLiteral;
import com.intellij.advancedExpressionFolding.processor.util.Consts;
import com.intellij.advancedExpressionFolding.settings.AdvancedExpressionFoldingSettings;
import com.intellij.psi.PsiLiteralExpression;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class LiteralExpressionExt {
    @Nullable
    public static Expression getLiteralExpression(@NotNull PsiLiteralExpression element) {
        if (element.getType() != null) {
            if (Consts.SUPPORTED_PRIMITIVE_TYPES.contains(element.getType().getCanonicalText())) {
                Object value = element.getValue();
                if (value instanceof Number) {
                    return new NumberLiteral(element, element.getTextRange(), null, (Number) value, false);
                } else if (value instanceof String) {
                    if (!element.isTextBlock()) {
                        return new StringLiteral(element, element.getTextRange(), (String) value);
                    } else if (AdvancedExpressionFoldingSettings.getInstance().getState().getLogFoldingTextBlocks()) {
                        return new StringLiteral(element, element.getTextRange(), (String) value);
                    }
                } else if (value instanceof Character) {
                    return new CharacterLiteral(element, element.getTextRange(), (Character) value);
                }
            }
        }
        return null;
    }
}

File: src/com/intellij/advancedExpressionFolding/processor/expression/LiteralExpressionExt.kt
package com.intellij.advancedExpressionFolding.processor.expression

import com.intellij.advancedExpressionFolding.expression.Expression
import com.intellij.advancedExpressionFolding.expression.literal.CharacterLiteral
import com.intellij.advancedExpressionFolding.expression.literal.NumberLiteral
import com.intellij.advancedExpressionFolding.expression.literal.StringLiteral
import com.intellij.advancedExpressionFolding.processor.util.Consts
import com.intellij.advancedExpressionFolding.settings.State
import com.intellij.advancedExpressionFolding.settings.state.ILogFoldingState
import com.intellij.psi.PsiLiteralExpression

object LiteralExpressionExt : ILogFoldingState by State()() {

    fun getLiteralExpression(element: PsiLiteralExpression): Expression? {
        val type = element.type ?: return null
        if (!Consts.SUPPORTED_PRIMITIVE_TYPES.contains(type.canonicalText)) {
            return null
        }
        val value = element.value
        return when (value) {
            is Number -> NumberLiteral(element, element.textRange, null, value, false)
            is String -> if (!element.isTextBlock || logFoldingTextBlocks) {
                StringLiteral(element, element.textRange, value)
            } else {
                null
            }
            is Char -> CharacterLiteral(element, element.textRange, value)
            else -> null
        }
    }
}
